File: src/Blokemon.Product/LocalProfile.cs
using System.Collections.Immutable;
using Blokemon.Core.SetDesign;

namespace Blokemon.Product;

public enum LocalProfileCreationFailure
{
    NoRegularCollectibleAvailable,
}

public enum PackOpenFailure
{
    ReceiptIdAlreadyUsed,
    ElevenCardPackUnavailable,
    AuthorityVersionMismatch,
    PackAllowanceExhausted,
}

public enum PackOpenDisposition
{
    Opened,
    AlreadyOpened,
}

public sealed class PackReceipt
{
    internal PackReceipt(
        PackReceiptId id,
        CommandId commandId,
        int sequence,
        ImmutableArray<CardId> sampledCollectibleIds
    )
    {
        Id = id;
        CommandId = commandId;
        Sequence = sequence;
        SampledCollectibleIds = sampledCollectibleIds;
    }

    public PackReceiptId Id { get; }

    public CommandId CommandId { get; }

    public int Sequence { get; }

    public ImmutableArray<CardId> SampledCollectibleIds { get; }
}

public sealed record PackOpenTransition(
    LocalProfile Profile,
    PackReceipt Receipt,
    PackOpenDisposition Disposition
);

public sealed partial class LocalProfile
{
    private readonly ImmutableDictionary<CardId, int> _collectibleOwnership;
    private readonly ImmutableDictionary<CommandId, PackReceipt> _receiptsByCommand;
    private readonly ImmutableDictionary<PackReceiptId, PackReceipt> _receiptsById;
    private readonly ImmutableDictionary<DeckId, SavedDeck> _savedDecks;
    private readonly ImmutableArray<StarterDeckClaim> _starterDeckClaims;

    private LocalProfile(
        ProfileId id,
        DisplayName displayName,
        string authorityManifestVersion,
        CardId guaranteedRegularCollectibleId,
        EconomyRules economy,
        ImmutableDictionary<CardId, int> collectibleOwnership,
        ImmutableDictionary<CommandId, PackReceipt> receiptsByCommand,
        ImmutableDictionary<PackReceiptId, PackReceipt> receiptsById,
        ImmutableDictionary<DeckId, SavedDeck> savedDecks,
        ImmutableArray<StarterDeckClaim> starterDeckClaims
    )
    {
        Id = id;
        DisplayName = displayName;
        BoundAuthorityManifestVersion = authorityManifestVersion;
        GuaranteedRegularCollectibleId = guaranteedRegularCollectibleId;
        Economy = economy;
        _collectibleOwnership = collectibleOwnership;
        _receiptsByCommand = receiptsByCommand;
        _receiptsById = receiptsById;
        _savedDecks = savedDecks;
        _starterDeckClaims = starterDeckClaims;
    }

    public ProfileId Id { get; }

    public DisplayName DisplayName { get; }

    public string BoundAuthorityManifestVersion { get; }

    public CardId GuaranteedRegularCollectibleId { get; }

    public EconomyRules Economy { get; }

    public int? RemainingPackAllowance =>
        EconomyRules.Remaining(Economy.PackAllowance, _receiptsById.Count);

    public int? RemainingStarterDeckClaimAllowance =>
        EconomyRules.Remaining(Economy.StarterDeckClaimAllowance, _starterDeckClaims.Length);

    public IReadOnlyDictionary<CardId, int> CollectibleOwnership => _collectibleOwnership;

    public IReadOnlyDictionary<PackReceiptId, PackReceipt> PackReceipts => _receiptsById;

    public IReadOnlyDictionary<DeckId, SavedDeck> SavedDecks => _savedDecks;

    public ImmutableArray<StarterDeckClaim> StarterDeckClaims => _starterDeckClaims;

    public StarterDeckClaim? LatestStarterDeckClaim =>
        _starterDeckClaims.IsEmpty ? null : _starterDeckClaims[^1];

    public static DomainResult<LocalProfile, LocalProfileCreationFailure> Create(
        ProfileId id,
        DisplayName displayName,
        BlokemonRuntimeManifest authority,
        EconomyRules? economy = null
    )
    {
        ArgumentNullException.ThrowIfNull(id);
        ArgumentNullException.ThrowIfNull(displayName);
        ArgumentNullException.ThrowIfNull(authority);

        var regular = authority
            .Collectibles.Where(static card => card.Rank == BlokemonRank.Regular)
            .OrderBy(static card => card.Id, StringComparer.Ordinal)
            .FirstOrDefault();
        if (regular is null)
        {
            return DomainResult<LocalProfile, LocalProfileCreationFailure>.Failure(
                LocalProfileCreationFailure.NoRegularCollectibleAvailable
            );
        }

        var regularId = CardId.FromAuthority(regular.Id);
        return DomainResult<LocalProfile, LocalProfileCreationFailure>.Success(
            new LocalProfile(
                id,
                displayName,
                authority.ManifestVersion,
                regularId,
                economy ?? EconomyRules.Unlimited,
                ImmutableDictionary<CardId, int>.Empty.Add(regularId, 1),
                ImmutableDictionary<CommandId, PackReceipt>.Empty,
                ImmutableDictionary<PackReceiptId, PackReceipt>.Empty,
                ImmutableDictionary<DeckId, SavedDeck>.Empty,
                ImmutableArray<StarterDeckClaim>.Empty
            )
        );
    }

    public int OwnedCollectibleQuantity(CardId cardId)
    {
        ArgumentNullException.ThrowIfNull(cardId);
        return _collectibleOwnership.GetValueOrDefault(cardId);
    }

    public DomainResult<PackOpenTransition, PackOpenFailure> OpenPack(
        CommandId commandId,
        PackReceiptId receiptId,
        BlokemonRuntimeManifest authority,
        IBlokemonRandomSource random
    )
    {
        ArgumentNullException.ThrowIfNull(commandId);
        ArgumentNullException.ThrowIfNull(receiptId);
        ArgumentNullException.ThrowIfNull(authority);
        ArgumentNullException.ThrowIfNull(random);

        if (_receiptsByCommand.TryGetValue(commandId, out var existingReceipt))
        {
            return DomainResult<PackOpenTransition, PackOpenFailure>.Success(
                new PackOpenTransition(this, existingReceipt, PackOpenDisposition.AlreadyOpened)
            );
        }

        if (
            !string.Equals(
                BoundAuthorityManifestVersion,
                authority.ManifestVersion,
                StringComparison.Ordinal
            )
        )
        {
            return DomainResult<PackOpenTransition, PackOpenFailure>.Failure(
                PackOpenFailure.AuthorityVersionMismatch
            );
        }

        if (_receiptsById.ContainsKey(receiptId))
        {
            return DomainResult<PackOpenTransition, PackOpenFailure>.Failure(
                PackOpenFailure.ReceiptIdAlreadyUsed
            );
        }

        if (Economy.PackAllowance is { } packAllowance && _receiptsById.Count >= packAllowance)
        {
            return DomainResult<PackOpenTransition, PackOpenFailure>.Failure(
                PackOpenFailure.PackAllowanceExhausted
            );
        }

        if (!CanSampleEleven(authority))
        {
            return DomainResult<PackOpenTransition, PackOpenFailure>.Failure(
                PackOpenFailure.ElevenCardPackUnavailable
            );
        }

        var sampledIds = BlokemonPackSampler
            .SampleEleven(authority, random)
            .Select(CardId.FromAuthority)
            .ToImmutableArray();
        var ownership = _collectibleOwnership;
        foreach (var cardId in sampledIds)
        {
            ownership = ownership.SetItem(cardId, ownership.GetValueOrDefault(cardId) + 1);
        }

        var receipt = new PackReceipt(receiptId, commandId, _receiptsById.Count + 1, sampledIds);
        var profile = Copy(
            collectibleOwnership: ownership,
            receiptsByCommand: _receiptsByCommand.Add(commandId, receipt),
            receiptsById: _receiptsById.Add(receiptId, receipt)
        );
        return DomainResult<PackOpenTransition, PackOpenFailure>.Success(
            new PackOpenTransition(profile, receipt, PackOpenDisposition.Opened)
        );
    }

    public DomainResult<DeckSaveTransition, DeckSaveFailure> CreateDeck(
        DeckId deckId,
        DeckName name,
        IEnumerable<DeckCardSelection> selections,
        BlokemonRuntimeManifest authority
    )
    {
        ArgumentNullException.ThrowIfNull(deckId);
        ArgumentNullException.ThrowIfNull(name);
        ArgumentNullException.ThrowIfNull(selections);
        ArgumentNullException.ThrowIfNull(authority);

        if (_savedDecks.ContainsKey(deckId))
        {
            return DomainResult<DeckSaveTransition, DeckSaveFailure>.Failure(
                new DeckSaveFailure.AlreadyExists(deckId)
            );
        }

        return DeckValidator
            .Validate(this, authority, selections)
            .Match(
                validDeck => SaveNewDeck(deckId, name, validDeck),
                issues =>
                    DomainResult<DeckSaveTransition, DeckSaveFailure>.Failure(
                        new DeckSaveFailure.InvalidDeck(issues)
                    )
            );
    }

    public DomainResult<DeckSaveTransition, DeckSaveFailure> ReviseDeck(
        DeckId deckId,
        DeckRevision expectedRevision,
        DeckName name,
        IEnumerable<DeckCardSelection> selections,
        BlokemonRuntimeManifest authority
    )
    {
        ArgumentNullException.ThrowIfNull(deckId);
        ArgumentNullException.ThrowIfNull(expectedRevision);
        ArgumentNullException.ThrowIfNull(name);
        ArgumentNullException.ThrowIfNull(selections);
        ArgumentNullException.ThrowIfNull(authority);

        if (!_savedDecks.TryGetValue(deckId, out var current))
        {
            return DomainResult<DeckSaveTransition, DeckSaveFailure>.Failure(
                new DeckSaveFailure.NotFound(deckId)
            );
        }

        if (current.Revision != expectedRevision)
        {
            return DomainResult<DeckSaveTransition, DeckSaveFailure>.Failure(
                new DeckSaveFailure.StaleRevision(deckId, expectedRevision, current.Revision)
            );
        }

        if (!current.Revision.TryNext(out var nextRevision))
        {
            return DomainResult<DeckSaveTransition, DeckSaveFailure>.Failure(
                new DeckSaveFailure.RevisionExhausted(deckId)
            );
        }

        return DeckValidator
            .Validate(this, authority, selections)
            .Match(
                validDeck => SaveRevisedDeck(deckId, name, nextRevision, validDeck),
                issues =>
                    DomainResult<DeckSaveTransition, DeckSaveFailure>.Failure(
                        new DeckSaveFailure.InvalidDeck(issues)
                    )
            );
    }

    // Deleting a deck removes only the deck. Collectible ownership, pack receipts and
    // starter claims are permanent history and are left exactly as they were.
    public DomainResult<DeckDeleteTransition, DeckDeleteFailure> DeleteDeck(DeckId deckId)
    {
        ArgumentNullException.ThrowIfNull(deckId);

        if (!_savedDecks.TryGetValue(deckId, out var deck))
        {
            return DomainResult<DeckDeleteTransition, DeckDeleteFailure>.Failure(
                DeckDeleteFailure.NotFound
            );
        }

        return DomainResult<DeckDeleteTransition, DeckDeleteFailure>.Success(
            new DeckDeleteTransition(Copy(savedDecks: _savedDecks.Remove(deckId)), deck)
        );
    }

    private DomainResult<DeckSaveTransition, DeckSaveFailure> SaveNewDeck(
        DeckId deckId,
        DeckName name,
        ValidatedDeck validDeck
    ) => SaveRevisedDeck(deckId, name, DeckRevision.Initial, validDeck);

    private DomainResult<DeckSaveTransition, DeckSaveFailure> SaveRevisedDeck(
        DeckId deckId,
        DeckName name,
        DeckRevision revision,
        ValidatedDeck validDeck
    )
    {
        var deck = new SavedDeck(deckId, name, revision, validDeck.Cards);
        var profile = Copy(savedDecks: _savedDecks.SetItem(deckId, deck));
        return DomainResult<DeckSaveTransition, DeckSaveFailure>.Success(
            new DeckSaveTransition(profile, deck)
        );
    }

    private static bool CanSampleEleven(BlokemonRuntimeManifest authority) =>
        authority.Products.Eleven.Count == 11
        && authority.Products.Eleven.Slots.Sum(static slot => (long)slot.Count) == 11
        && authority.Products.Eleven.Slots.All(slot =>
            slot.Count >= 0
            && authority.Collectibles.Count(card => card.ProductBucket == slot.Bucket) >= slot.Count
        );

    private LocalProfile Copy(
        ImmutableDictionary<CardId, int>? collectibleOwnership = null,
        ImmutableDictionary<CommandId, PackReceipt>? receiptsByCommand = null,
        ImmutableDictionary<PackReceiptId, PackReceipt>? receiptsById = null,
        ImmutableDictionary<DeckId, SavedDeck>? savedDecks = null,
        ImmutableArray<StarterDeckClaim>? starterDeckClaims = null
    ) =>
        new(
            Id,
            DisplayName,
            BoundAuthorityManifestVersion,
            GuaranteedRegularCollectibleId,
            Economy,
            collectibleOwnership ?? _collectibleOwnership,
            receiptsByCommand ?? _receiptsByCommand,
            receiptsById ?? _receiptsById,
            savedDecks ?? _savedDecks,
            starterDeckClaims ?? _starterDeckClaims
        );
}

File: src/Blokemon.Product/LocalProfile.fs
namespace rec Blokemon.Product

open System
open System.Collections.Generic
open System.Collections.Immutable
open Blokemon.Core.SetDesign

/// The profile that opening a pack produced, with the receipt it wrote.
type PackOpenTransition =
    { Profile: LocalProfile
      Receipt: PackReceipt
      Disposition: PackOpenDisposition }

/// The profile that saving a deck produced, with the deck it wrote.
type DeckSaveTransition =
    { Profile: LocalProfile
      Deck: SavedDeck }

/// The profile that deleting a deck produced, with the deck it removed.
type DeckDeleteTransition =
    { Profile: LocalProfile
      Deck: SavedDeck }

/// The outcome of claiming a starter deck.
[<RequireQualifiedAccess>]
type StarterDeckClaimOutcome =
    | Claimed of Profile: LocalProfile * Claim: StarterDeckClaim
    | AlreadyClaimed of Profile: LocalProfile * Claim: StarterDeckClaim

// The type carries the whole of a profile and none of the work: every operation below hands
// its state to the module that owns that concern and wraps whatever comes back.
/// One machine's local player: what it owns, what it opened, and what it saved.
type LocalProfile =
    private
        { state: LocalProfileState }

    member this.Id = this.state.id

    member this.DisplayName = this.state.displayName

    member this.BoundAuthorityManifestVersion = this.state.boundAuthorityManifestVersion

    member this.GuaranteedRegularCollectibleId = this.state.guaranteedRegularCollectibleId

    member this.Economy = this.state.economy

    member this.RemainingPackAllowance =
        EconomyRules.Remaining(this.state.economy.PackAllowance, this.state.receiptsById.Count)

    member this.RemainingStarterDeckClaimAllowance =
        EconomyRules.Remaining(
            this.state.economy.StarterDeckClaimAllowance,
            this.state.starterDeckClaims.Length
        )

    member this.CollectibleOwnership: IReadOnlyDictionary<CardId, int> =
        this.state.collectibleOwnership

    member this.PackReceipts: IReadOnlyDictionary<PackReceiptId, PackReceipt> =
        this.state.receiptsById

    member this.SavedDecks: IReadOnlyDictionary<DeckId, SavedDeck> = this.state.savedDecks

    member this.StarterDeckClaims = this.state.starterDeckClaims

    member this.LatestStarterDeckClaim: StarterDeckClaim | null =
        this.state.LatestStarterDeckClaim

    /// A new profile bound to the current authority, playing under unlimited rules.
    static member Create
        (id: ProfileId, displayName: DisplayName, authority: BlokemonRuntimeManifest)
        =
        LocalProfile.Create(id, displayName, authority, EconomyRules.Unlimited)

    /// A new profile bound to the current authority.
    static member Create
        (
            id: ProfileId,
            displayName: DisplayName,
            authority: BlokemonRuntimeManifest,
            economy: EconomyRules | null
        ) =
        ArgumentNullException.ThrowIfNull(id, nameof id)
        ArgumentNullException.ThrowIfNull(displayName, nameof displayName)
        ArgumentNullException.ThrowIfNull(authority, nameof authority)

        let regular =
            authority.Collectibles
            |> Array.filter (fun card -> card.Rank = BlokemonRank.Regular)
            |> Array.sortWith (fun left right -> String.CompareOrdinal(left.Id, right.Id))
            |> Array.tryHead

        match regular with
        | None -> DomainResult.Failed LocalProfileCreationFailure.NoRegularCollectibleAvailable
        | Some card ->
            let regularId = CardId.FromAuthority card.Id

            DomainResult.Succeeded
                { state =
                    { id = id
                      displayName = displayName
                      boundAuthorityManifestVersion = authority.ManifestVersion
                      historicalAuthorityManifestVersions = ImmutableArray<string>.Empty
                      unavailableHistoricalCardIds = Set.empty
                      guaranteedRegularCollectibleId = regularId
                      economy =
                        match economy with
                        | null -> EconomyRules.Unlimited
                        | rules -> rules
                      collectibleOwnership =
                        ImmutableDictionary<CardId, int>.Empty.Add(regularId, 1)
                      receiptsByCommand = ImmutableDictionary<CommandId, PackReceipt>.Empty
                      receiptsById = ImmutableDictionary<PackReceiptId, PackReceipt>.Empty
                      savedDecks = ImmutableDictionary<DeckId, SavedDeck>.Empty
                      starterDeckClaims = ImmutableArray<StarterDeckClaim>.Empty } }

    /// How many copies of a card the profile owns.
    member this.OwnedCollectibleQuantity(cardId: CardId) =
        ArgumentNullException.ThrowIfNull(cardId, nameof cardId)
        this.state.OwnedQuantity cardId

    /// Rebinds a restored profile to the checked-out authority without changing its product state.
    member this.MigrateAuthority(currentAuthority: BlokemonRuntimeManifest) =
        ArgumentNullException.ThrowIfNull(currentAuthority, nameof currentAuthority)

        if
            String.Equals(
                this.state.boundAuthorityManifestVersion,
                currentAuthority.ManifestVersion,
                StringComparison.Ordinal
            )
        then
            this
        else
            let currentCardIds =
                Seq.concat
                    [ currentAuthority.Collectibles |> Seq.map _.Id
                      currentAuthority.Kits |> Seq.map _.Id
                      currentAuthority.BasicVim |> Seq.map _.Id ]
                |> Set.ofSeq

            let unavailableCardIds =
                seq {
                    yield this.state.guaranteedRegularCollectibleId
                    yield! this.state.collectibleOwnership.Keys

                    for receipt in this.state.receiptsById.Values do
                        yield! receipt.SampledCollectibleIds

                    for deck in this.state.savedDecks.Values do
                        yield! deck.Cards.Keys

                    for claim in this.state.starterDeckClaims do
                        yield! claim.CollectibleGrants |> Seq.map _.CardId
                }
                |> Seq.filter (fun cardId -> not (currentCardIds.Contains cardId.Value))
                |> Seq.fold
                    (fun historical cardId -> historical |> Set.add cardId)
                    this.state.unavailableHistoricalCardIds

            { state =
                { this.state with
                    boundAuthorityManifestVersion = currentAuthority.ManifestVersion
                    historicalAuthorityManifestVersions =
                        if
                            this.state.historicalAuthorityManifestVersions.Contains(
                                this.state.boundAuthorityManifestVersion
                            )
                        then
                            this.state.historicalAuthorityManifestVersions
                        else
                            this.state.historicalAuthorityManifestVersions.Add(
                                this.state.boundAuthorityManifestVersion
                            )
                    unavailableHistoricalCardIds = unavailableCardIds } }

    /// Opens an eleven-card pack, replaying a saved receipt when the command repeats.
    member this.OpenPack
        (
            commandId: CommandId,
            receiptId: PackReceiptId,
            authority: BlokemonRuntimeManifest,
            random: IBlokemonRandomSource
        ) : DomainResult<PackOpenTransition, PackOpenFailure> =
        match ProfilePacks.openPack this.state commandId receiptId authority random with
        | DomainResult.Failed failure -> DomainResult.Failed failure
        | DomainResult.Succeeded(ReplayedPack receipt) ->
            DomainResult.Succeeded
                { Profile = this
                  Receipt = receipt
                  Disposition = PackOpenDisposition.AlreadyOpened }
        | DomainResult.Succeeded(DrewPack(opened, receipt)) ->
            DomainResult.Succeeded
                { Profile = { state = opened }
                  Receipt = receipt
                  Disposition = PackOpenDisposition.Opened }

    /// Saves a new deck.
    member this.CreateDeck
        (
            deckId: DeckId,
            name: DeckName,
            selections: IEnumerable<DeckCardSelection>,
            authority: BlokemonRuntimeManifest
        ) : DomainResult<DeckSaveTransition, DeckSaveFailure> =
        match ProfileDecks.createDeck this.state deckId name selections authority with
        | DomainResult.Failed failure -> DomainResult.Failed failure
        | DomainResult.Succeeded(saved, deck) ->
            DomainResult.Succeeded
                { Profile = { state = saved }
                  Deck = deck }

    /// Replaces a saved deck, provided the caller holds its current revision.
    member this.ReviseDeck
        (
            deckId: DeckId,
            expectedRevision: DeckRevision,
            name: DeckName,
            selections: IEnumerable<DeckCardSelection>,
            authority: BlokemonRuntimeManifest
        ) : DomainResult<DeckSaveTransition, DeckSaveFailure> =
        match
            ProfileDecks.reviseDeck this.state deckId expectedRevision name selections authority
        with
        | DomainResult.Failed failure -> DomainResult.Failed failure
        | DomainResult.Succeeded(saved, deck) ->
            DomainResult.Succeeded
                { Profile = { state = saved }
                  Deck = deck }

    /// Removes a saved deck.
    member this.DeleteDeck(deckId: DeckId) : DomainResult<DeckDeleteTransition, DeckDeleteFailure> =
        match ProfileDecks.deleteDeck this.state deckId with
        | DomainResult.Failed failure -> DomainResult.Failed failure
        | DomainResult.Succeeded(remaining, deck) ->
            DomainResult.Succeeded
                { Profile = { state = remaining }
                  Deck = deck }

    /// Claims a starter deck, granting its full collectible contents and saving its deck.
    member this.ClaimStarterDeck
        (
            commandId: CommandId,
            definition: StarterDeckDefinition,
            currentAuthority: BlokemonRuntimeManifest
        ) : DomainResult<StarterDeckClaimOutcome, StarterDeckClaimFailure> =
        match ProfileClaims.claimStarterDeck this.state commandId definition currentAuthority with
        | DomainResult.Failed failure -> DomainResult.Failed failure
        | DomainResult.Succeeded(AlreadyHeldStarter claim) ->
            DomainResult.Succeeded(StarterDeckClaimOutcome.AlreadyClaimed(this, claim))
        | DomainResult.Succeeded(GrantedStarter(granted, claim)) ->
            DomainResult.Succeeded(StarterDeckClaimOutcome.Claimed({ state = granted }, claim))

    /// The persisted form of the profile.
    member this.ToSnapshot() : LocalProfileSnapshot =
        ProfileSnapshotProjection.toSnapshot this.state

    /// The profile a persisted snapshot describes, or why it is not one.
    static member Restore
        (snapshot: LocalProfileSnapshot, currentAuthority: BlokemonRuntimeManifest)
        : DomainResult<LocalProfile, LocalProfileRestorationFailure> =
        match ProfileRestoration.restore snapshot currentAuthority with
        | DomainResult.Failed failure -> DomainResult.Failed failure
        | DomainResult.Succeeded restored -> DomainResult.Succeeded { state = restored }
